src/toxcore/net_crypto/errors.rs

```diff
@@ -0,0 +1,612 @@
+        self.ctx.fmt(f)
+    }
+}
+
+/// The specific kind of error that can occur.
+#[derive(Debug, Eq, PartialEq, Fail)]
+pub enum SendDataErrorKind {
+    /// Connection is not established.
+    #[fail(display = "Connection is not established")]
+    NoConnection,
+    /// Error indicates that sending response packet error.
+    #[fail(display = "Sending response error")]
+    SendTo,
+}
+
+impl From<SendDataErrorKind> for SendDataError {
+    fn from(kind: SendDataErrorKind) -> SendDataError {
+        SendDataError::from(Context::new(kind))
+    }
+}
+
+impl From<Context<SendDataErrorKind>> for SendDataError {
+    fn from(ctx: Context<SendDataErrorKind>) -> SendDataError {
+        SendDataError { ctx }
+    }
+}
+
+impl From<TimeoutError<mpsc::SendError<(Packet, SocketAddr)>>> for SendDataError {
+    fn from(error: TimeoutError<mpsc::SendError<(Packet, SocketAddr)>>) -> SendDataError {
+        SendDataError {
+            ctx: error.context(SendDataErrorKind::SendTo)
+        }
+    }
+}
+
+/// Error that can happen when calling `run`.
+#[derive(Debug)]
+pub struct RunError {
+    ctx: Context<RunErrorKind>,
+}
+
+impl RunError {
+    /// Return the kind of this error.
+    pub fn kind(&self) -> &RunErrorKind {
+        self.ctx.get_context()
+    }
+}
+
+impl Fail for RunError {
+    fn cause(&self) -> Option<&Fail> {
+        self.ctx.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.ctx.backtrace()
+    }
+}
+
+impl fmt::Display for RunError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.ctx.fmt(f)
+    }
+}
+
+/// The specific kind of error that can occur.
+#[derive(Clone, Debug, Eq, PartialEq, Fail)]
+pub enum RunErrorKind {
+    /// Sending pings error
+    #[fail(display = "Sending crypto data packet error")]
+    SendData,
+    /// NetCrypto periodical wakeup timer error
+    #[fail(display = "Netcrypto periodical wakeup timer error")]
+    Wakeup,
+}
+
+impl From<RunErrorKind> for RunError {
+    fn from(kind: RunErrorKind) -> RunError {
+        RunError::from(Context::new(kind))
+    }
+}
+
+impl From<Context<RunErrorKind>> for RunError {
+    fn from(ctx: Context<RunErrorKind>) -> RunError {
+        RunError { ctx }
+    }
+}
+
+impl From<TimerError> for RunError {
+    fn from(error: TimerError) -> RunError {
+        RunError {
+            ctx: error.context(RunErrorKind::Wakeup)
+        }
+    }
+}
+
+impl From<SendDataError> for RunError {
+    fn from(error: SendDataError) -> RunError {
+        RunError {
+            ctx: error.context(RunErrorKind::SendData)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::toxcore::io_tokio::*;
+
+    use std::time::Duration;
+    use futures::future::Future;
+
+    #[test]
+    fn packets_array_error() {
+        let error = PacketsArrayError::from(PacketsArrayErrorKind::ArrayFull);
+        assert!(error.cause().is_none());
+        assert!(error.backtrace().is_some());
```

Comment:
Isn't bactrace optional?

Reply:
Yes. But we introduced `backtrace` to find the location of error.
I think we should enforce to use `backtrace`.

Reply:
No, backtrace is optional